Fallback to regex for dynamic like pattern

gaurav8297 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

The compilation time for DFA is costly compared to regex. Therefore, for dynamic like pattern it is better
to use a regex-based matcher.

## Benchmarks

Before:

query: select count(shipmode) from lineitem where shipinstruct like concat('%', concat("shipmode", '%'));
time: 27:04 mins

After:

query: select count(shipmode) from lineitem where shipinstruct like concat('%', concat("shipmode", '%'));
time: 4:44 sec

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

lukasz-stec commented on core/trino-main/src/main/java/io/trino/type/LikePatternType.java:
```diff
@@ -71,7 +72,7 @@ public Object getObject(Block block, int position)
         }
 
-        return LikeMatcher.compile(pattern, escape);
+        return DfaLikeMatcher.compile(pattern, escape);
```
I think you have to support both matchers here because this is used to serialize function call via `io.trino.sql.planner.LiteralEncoder#toExpression`

lukasz-stec reviewed (commented):
general approach lgtm.
One thing to add is a query test with a dynamic like pattern, possibly to `io.trino.testing.BaseConnectorTest#testLikePredicate`

lukasz-stec commented on core/trino-main/src/test/java/io/trino/operator/scalar/BenchmarkLike.java:
```diff
@@ -91,7 +91,7 @@
         private byte[] bytes;
         private JoniRegexp joniPattern;
-        private LikeMatcher matcher;
+        private DfaLikeMatcher matcher;
```
Do you think we also need a benchmark for RegexpLikeMatcher?

lukasz-stec commented on core/trino-main/src/test/java/io/trino/type/TestLikePatternType.java:
```diff
@@ -30,15 +30,15 @@
         assertThat(pattern.getEscape()).isEqualTo(Optional.of('X'));
 
-        pattern = (LikeMatcher) LIKE_PATTERN.getObject(block, 1);
+        pattern = (DfaLikeMatcher) LIKE_PATTERN.getObject(block, 1);
```
We also need tests for RegexpLikeMatcher

lukasz-stec reviewed (approved):
lgtm % small comments

martint commented:
The are probably opportunities for improving compilation time for the DFA-based like matcher. I would start with analyzing where the performance is going before we add and alternate implementation based on Joni.

gaurav8297 commented:
> There are probably opportunities for improving compilation time for the DFA-based like matcher. I would start with analyzing where the performance is going before we add and alternate implementation based on Joni.

The majority of time and CPU is spent during the conversion of NFA to DFA. The time complexity for this conversion is O(2^n) where n is the number of states in NFA. On the other hand for matching a string using DFA it is just O(n). The compilation time complexity is higher because we need to look at all the subsets of NFA states (2^n) to create a DFA. Therefore, IMO DFA-based like matcher is a good choice if matching is more frequent compared to compilation. However, I still tried to improve performance and I've mentioned my process below.

### Current Performance and Bottlenecks:

**Compilation:**

```
BenchmarkLikePattern.testCurrentDFA   abc%def%ghi  avgt   20  1577020.207 ± 1974381.187  ns/op
BenchmarkLikePattern.testRegex        abc%def%ghi  avgt   20     2239.984 ±     593.622  ns/op
```
**Matching:**

```
BenchmarkLikePattern.testCurrentDFA   abc%def%ghi  avgt   10   71.252 ±  7.339  ns/op
BenchmarkLikePattern.testRegex        abc%def%ghi  avgt   10  381.028 ± 23.259  ns/op
```

Most of the time is spent on `NFA.toDFA()`. Please take a look at the JFR profile.

<img width="1333" alt="Screenshot 2023-01-16 at 1 34 49 PM" src="https://user-images.githubusercontent.com/20029138/212673866-cf7e55c6-5377-4b12-9dac-e9ad62ed6629.png">

<img width="1355" alt="Screenshot 2023-01-16 at 1 33 54 PM" src="https://user-images.githubusercontent.com/20029138/212673889-fadd00c1-4df3-48dd-a1f8-ba3e6e5bdeff.png">


### Experiment:
I tried to improve performance without changing any core logic. I just used type-specific maps, sets and queues from `fastutil` and removed extra immutable copy operations. It gives a considerable bump in performance but is still far from the baseline.

Results:

```
BenchmarkLikePattern.testCurrentDFA         avgt   20  1006924.581 ± 414884.827  ns/op
BenchmarkLikePattern.testOptimizedDFA       avgt   20   336407.127 ± 109607.127  ns/op
BenchmarkLikePattern.testRegex              avgt   20     1611.863 ±    229.578  ns/op
```

Next, I tried to change the core logic in the following way
- First, I removed the concept of failed states from the final DFA considering we can handle it during match time using the `transitions` array in `DenseDfaMatcher`.
- Second, I tried to remove the `Prefix` pattern which is used for a UTF-8 state machine. I thought it kind of increases the number of nodes in the state machine so removing that might give performance gains. Instead, I added a transition called `UTF8` in the final DFA. However, I still have to figure out how to handle this as a special condition inside `DenseDfaMatcher` or if at all it is possible. Additionally, it could have a major impact on performance during match time.

Even after these changes, I couldn't achieve results close to Joni regex.

```
BenchmarkLikePattern.testCurrentDFA         avgt   20  1005057.695 ± 684952.614  ns/op
BenchmarkLikePattern.testOptimizedDFA       avgt   20     3316.518 ±   4443.102  ns/op
BenchmarkLikePattern.testRegex              avgt   20     1608.885 ±    202.779  ns/op
````

<br>

**Final comment:** Considering matching + compilation time even after these experiments, it makes sense to go with Joni regex for dynamic like pattern.

Also, please let me know if you have some more suggestions which I can try.

@martint @lukasz-stec @sopel39

Dith3r commented on core/trino-main/src/test/java/io/trino/likematcher/TestLikeMatcher.java:
```diff
@@ -24,59 +25,65 @@
     {
         // min length short-circuit
-        assertFalse(match("__", "a"));
+        assertFalse(match(type, "__", "a"));
```
nit. No multiline string?

martint commented:
See https://github.com/trinodb/trino/pull/15999
